### PR TITLE
Assets - Extract default error catalog, normalize constants (#895)

### LINKED ISSUE #895: Assets – Default Error Catalog Extraction and Constants Normalization
# Technical Requirements Document: Assets – Default Error Catalog Extraction and Constants Normalization

**Project:** Tiferet Framework
**Repository:** https://github.com/greatstrength/tiferet
**Date:** July 3, 2026
**Version:** Core DDD Parity III – Domain, Assets, and Mapper Parity (tracking milestone)

## 1. Overview

In `main`, `tiferet/assets/constants.py` mixes two distinct concerns: the framework's error-code identifier constants and a large, hand-maintained `DEFAULT_ERRORS` dictionary whose entries are inline dict literals. The current file also carries duplicated identifier declarations, an ungrouped block of utility error IDs, raw string-literal keys for several catalog entries, and a typo in the module docstring. This story separates the two concerns: identifier constants plus a small `create_default_error` factory and the `EN_US` locale constant remain in `constants.py`, while the default error catalog moves into a dedicated `tiferet/assets/error.py` module that builds each definition through the factory. It also normalizes the identifier constants (one artifact-commented declaration each, no duplicates, named constants for every catalog key) and adds the first unit test coverage for the assets layer. The public `DEFAULT_ERRORS` export is preserved through `tiferet/assets/__init__.py`, so consumers (`ErrorContext`, error domain events) are unaffected.

### In Scope
- Fix the `constants.py` module docstring and add typing imports.
- Add the `EN_US` locale constant to `constants.py`.
- Add the `create_default_error` factory function to `constants.py`.
- Normalize the error-code identifier constants in `constants.py`: deduplicate, give each a `# ** constant:` artifact comment, and promote every catalog key to a named `*_ID` constant.
- Remove the inline `DEFAULT_ERRORS` dictionary from `constants.py`.
- Add a new `tiferet/assets/error.py` module that defines each default error via `create_default_error` and assembles `DEFAULT_ERRORS`.
- Re-point the `DEFAULT_ERRORS` re-export in `tiferet/assets/__init__.py` from `.constants` to `.error`.
- Add `tests/assets/test_constants.py` covering `EN_US` and `create_default_error`.

### Out of Scope
- Any change to `TiferetError` / `TiferetAPIError` (`assets/exceptions.py`).
- Error resolution/formatting behavior in `ErrorContext` or the error domain events.
- Message-template wording changes beyond aligning placeholders with the runtime call sites already present in `main`.
- Localization beyond the existing `en_US` messages.

## 3. Components Affected

| Component | File/Path | Changes |
|-----------|-----------|---------|
| Assets constants | `tiferet/assets/constants.py` | Update docstring + imports; Add `EN_US`; Add `create_default_error`; Update/normalize error-ID constants; Remove inline `DEFAULT_ERRORS` |
| Assets error catalog | `tiferet/assets/error.py` | Add new module: default error definitions built via `create_default_error`; assemble `DEFAULT_ERRORS` |
| Assets exports | `tiferet/assets/__init__.py` | Update `DEFAULT_ERRORS` import source from `.constants` to `.error` (public API preserved) |
| Assets constants tests | `tests/assets/test_constants.py` | Add new test module covering `EN_US` and `create_default_error` |
| Assets tests package | `tests/assets/__init__.py` | Add package marker for the new test package |

### 4.1 `constants.py` — module docstring and imports

- Correct the module docstring to `"""Tiferet Constants (Assets)"""` (fixes the existing `Connstants` typo).
- Add a `# *** imports` / `# ** core` section importing `List, Tuple, Dict, Any` from `typing` (required by the factory signature).

### 4.2 `constants.py` — `EN_US` locale constant

Add a general `# *** constants` section preceding the error section:

```python
# *** constants

# ** constant: en_us
EN_US = 'en_US'
```

- Value: `'en_US'`.
- Purpose: single source of truth for the default locale code, consumed by `error.py` message pairs and available to consumers via `a.const.EN_US`.

### 4.3 `constants.py` — error-code identifier normalization

Consolidate all error-code identifiers under a single `# *** constants (error)` section. Each identifier is declared exactly once with its own `# ** constant: <snake_case>` artifact comment and a `NAME_ID = 'NAME'` string value.

Normalization rules (Add / Update / Remove):

| From (current `main`) | To (target) | Action |
|-----------------------|-------------|--------|
| Duplicate declarations of `FEATURE_NAME_REQUIRED_ID`, `INVALID_FEATURE_ATTRIBUTE_ID`, `INVALID_MODEL_ATTRIBUTE_ID`, `INVALID_APP_INTERFACE_TYPE_ID` | Single declaration each | Update (dedupe) |
| Ungrouped aligned block under `# ** constants: utils_errors` (YAML/JSON/CSV/TOML IDs) | Individual `# ** constant:` declarations in the unified section | Update |
| Raw string keys in the catalog: `'APP_ERROR'`, `'CONFIG_FILE_NOT_FOUND'`, `'APP_CONFIG_LOADING_FAILED'`, `'CONTAINER_CONFIG_LOADING_FAILED'`, `'FEATURE_CONFIG_LOADING_FAILED'`, `'ERROR_CONFIG_LOADING_FAILED'`, `'CLI_CONFIG_LOADING_FAILED'` | Named constants `APP_ERROR_ID`, `CONFIG_FILE_NOT_FOUND_ID`, `APP_CONFIG_LOADING_FAILED_ID`, `CONTAINER_CONFIG_LOADING_FAILED_ID`, `FEATURE_CONFIG_LOADING_FAILED_ID`, `ERROR_CONFIG_LOADING_FAILED_ID`, `CLI_CONFIG_LOADING_FAILED_ID` | Add |
| (absent) | `CONTEXT_NOT_FOUND_ID = 'CONTEXT_NOT_FOUND'` | Add |

- The resulting section defines one identifier per framework error code (73 identifiers total), each following the `# ** constant:` convention.
- Identifier string values must remain stable for codes that already exist in `main` (e.g., `FEATURE_NOT_FOUND_ID == 'FEATURE_NOT_FOUND'`), since they are matched against `TiferetError.error_code` at runtime.

### 4.4 `constants.py` — `create_default_error` factory

Add a `# *** functions` section with a single factory:

```python
# *** functions

# ** function: create_default_error
def create_default_error(id: str, name: str, messages: List[Tuple[str, str]]) -> Dict[str, Any]:
    '''
    Build a default error definition dictionary.

    :param id: The unique identifier of the error.
    :type id: str
    :param name: The human-readable error name.
    :type name: str
    :param messages: Ordered (lang, text) message pairs.
    :type messages: List[Tuple[str, str]]
    :return: The default error definition.
    :rtype: Dict[str, Any]
    '''

    # Assemble and return the default error definition dictionary.
    return {
        'id': id,
        'name': name,
        'message': [{'lang': lang, 'text': text} for lang, text in messages],
    }
```

- Return shape: `{'id': <id>, 'name': <name>, 'message': [{'lang': <lang>, 'text': <text>}, ...]}`.
- Message order is preserved from the supplied `messages` sequence.
- An empty `messages` list yields `'message': []`.

### 4.5 `constants.py` — remove inline `DEFAULT_ERRORS`

- Remove the `DEFAULT_ERRORS` dictionary and its inline `# * error:` entries from `constants.py`. The catalog now lives in `error.py` (§4.6). `constants.py` no longer defines `DEFAULT_ERRORS`.

### 4.6 `error.py` — new default error catalog module

Add `tiferet/assets/error.py`:

- **Module docstring:** describe the module as the default error catalog mapping each error-code constant to its name and multilingual message templates, consumed by `ErrorContext` and the error domain events when not overridden by consumer configuration.
- **Imports (`# *** imports` / `# ** app`):** import `EN_US`, `create_default_error`, and every `*_ID` constant from `.constants` (alphabetically ordered).
- **Definitions (`# *** constants`):** one module-level constant per error, each with a `# ** constant: <snake_case>` comment, built via `create_default_error(<ID>, <Name>, [(EN_US, <text>)])`. Example:

```python
# ** constant: error_not_found
ERROR_NOT_FOUND = create_default_error(
    ERROR_NOT_FOUND_ID,
    'Error Not Found',
    [(EN_US, 'Error not found: {id}.')],
)
```

- **Catalog assembly (`# ** constant: default_errors`):** a `DEFAULT_ERRORS` dict mapping each `*_ID` constant to its corresponding definition, one entry per identifier defined in §4.3 (73 entries).
- Message placeholders must match the keyword arguments passed at the runtime raise sites already present in `main` (e.g., `FEATURE_COMMAND_LOADING_FAILED` uses `{service_id}`, `DEPENDENCY_TYPE_NOT_FOUND` uses `{configuration_id}` and `{flags}`).

### 4.7 `assets/__init__.py` — export source change

- Import `DEFAULT_ERRORS` from `.error` instead of `.constants`; continue exporting `ERROR_NOT_FOUND_ID` and `const` from `.constants`.
- `DEFAULT_ERRORS` and `ERROR_NOT_FOUND_ID` remain importable as `from tiferet.assets import DEFAULT_ERRORS` and via `a.DEFAULT_ERRORS`, preserving the public API used across the framework.

### 4.8 Test requirements

Add `tests/assets/__init__.py` (package marker) and `tests/assets/test_constants.py` with standalone tests following the `# *** imports` / `# *** tests` structure:

| Test name | What it verifies |
|-----------|-----------------|
| `test_en_us_constant` | `EN_US == 'en_US'` |
| `test_create_default_error_single_message` | Returns `{'id', 'name', 'message': [{'lang': 'en_US', 'text': ...}]}` for a single pair |
| `test_create_default_error_preserves_message_order` | Multiple `(lang, text)` pairs are emitted in the supplied order |
| `test_create_default_error_empty_messages` | An empty `messages` list yields `'message': []` while `id`/`name` are retained |

## 5. Acceptance Criteria

1. `tiferet/assets/constants.py` module docstring reads `"""Tiferet Constants (Assets)"""` and no longer contains the `DEFAULT_ERRORS` dictionary.
2. `EN_US == 'en_US'` is importable from `tiferet.assets.constants` and via `a.const.EN_US`.
3. `create_default_error('X', 'X Name', [(EN_US, 'msg {a}.')])` returns `{'id': 'X', 'name': 'X Name', 'message': [{'lang': 'en_US', 'text': 'msg {a}.'}]}`; message order is preserved and an empty list yields `'message': []`.
4. Every error-code identifier is declared exactly once with a `# ** constant:` comment; no duplicate declarations remain; `CONTEXT_NOT_FOUND_ID` and the seven previously string-keyed identifiers (`APP_ERROR_ID`, `CONFIG_FILE_NOT_FOUND_ID`, `APP_CONFIG_LOADING_FAILED_ID`, `CONTAINER_CONFIG_LOADING_FAILED_ID`, `FEATURE_CONFIG_LOADING_FAILED_ID`, `ERROR_CONFIG_LOADING_FAILED_ID`, `CLI_CONFIG_LOADING_FAILED_ID`) exist as named constants.
5. `tiferet/assets/error.py` exists, builds every definition through `create_default_error`, and `DEFAULT_ERRORS` contains one entry per error-code constant (73 entries), keyed by the `*_ID` constants.
6. `from tiferet.assets import DEFAULT_ERRORS` and `a.DEFAULT_ERRORS` resolve to the catalog defined in `error.py`.
7. `tests/assets/test_constants.py` exists and passes (`pytest tests/assets/test_constants.py`).
8. The full suite passes (`pytest tiferet/ tests/`) with no regressions in `ErrorContext` or error-event tests.

## 6. Non-Functional Requirements

- All new/updated code follows the structured code style: `# ***` / `# **` / `# *` artifact comments, RST docstrings, and the standard spacing rules.
- The change is backward-compatible: the public `DEFAULT_ERRORS` and `ERROR_NOT_FOUND_ID` symbols keep their import paths through `tiferet.assets`.
- Identifier string values for pre-existing error codes are unchanged, preserving runtime `error_code` matching.
- Separation of concerns: `constants.py` holds identifiers + factory + locale; `error.py` holds catalog data — improving readability, testability, and future localization.
- `error.py` imports are alphabetically ordered for maintainability.

## 7. Prerequisites

| Dependency | Status in `main` |
|------------|-----------------|
| `TiferetError` / `TiferetAPIError` (`assets/exceptions.py`) | Present — unchanged by this story |
| `DEFAULT_ERRORS` consumers (`ErrorContext`, error domain events) resolve the catalog via the `tiferet.assets` package | Present — continue to work via the preserved re-export |
| `tests/` top-level test tree | Present — new `tests/assets/` package added under it |

## 8. Related Code Style Documentation

- `tiferet-code-style` / [code_style.md](https://github.com/greatstrength/tiferet/blob/main/docs/core/code_style.md)
- `tiferet-code-assets` / [assets.md](https://github.com/greatstrength/tiferet/blob/main/docs/core/assets.md)

### PR DESCRIPTION
**Pull request by Ashatz**
## Summary

Resolves #895 — Assets: Default Error Catalog Extraction and Constants Normalization (Parity III Story 1, P0).

Separates two concerns that were conflated in `tiferet/assets/constants.py`: error-code identifier constants (which stay in `constants.py`) and the default error catalog (which moves to a new `tiferet/assets/error.py`). Also normalizes all identifier constants and adds the first unit test coverage for the assets layer.

### `tiferet/assets/constants.py`
- Fixed module docstring typo (`Connstants` → `Constants`)
- Added `# *** imports` section with `typing` imports required by the factory
- Added `EN_US = 'en_US'` locale constant under `# *** constants`
- Deduplicated and normalized all 73 error-code identifier constants under a single `# *** constants (error)` section; each has its own `# ** constant:` artifact comment
- Added 8 new named `*_ID` constants for previously raw-string catalog keys: `APP_ERROR_ID`, `CONFIG_FILE_NOT_FOUND_ID`, `APP_CONFIG_LOADING_FAILED_ID`, `CONTAINER_CONFIG_LOADING_FAILED_ID`, `FEATURE_CONFIG_LOADING_FAILED_ID`, `ERROR_CONFIG_LOADING_FAILED_ID`, `CLI_CONFIG_LOADING_FAILED_ID`
- Added `CONTEXT_NOT_FOUND_ID = 'CONTEXT_NOT_FOUND'`
- Added `create_default_error` factory function under `# *** functions`
- Removed the inline `DEFAULT_ERRORS` dictionary

### `tiferet/assets/error.py` (new)
- Imports all 73 `*_ID` constants, `EN_US`, and `create_default_error` from `.constants` (alphabetically ordered)
- Defines one named constant per error (e.g., `FEATURE_NOT_FOUND`) built via `create_default_error`
- Assembles `DEFAULT_ERRORS` dict with 73 entries keyed by `*_ID` constants

### `tiferet/assets/__init__.py`
- Re-points `DEFAULT_ERRORS` import from `.constants` to `.error`; public API preserved

### `tests/assets/` (new)
- `__init__.py` package marker
- `test_constants.py`: 4 unit tests covering `EN_US`, single-message factory output, message-order preservation, and empty-messages edge case

## Test Results
- 4 new tests: all pass
- 760 existing tests: all pass, 11 skipped (pre-existing skips)
- The 8 pre-existing import errors in `tiferet/contexts/tests/` and `tiferet/blueprints/tests/` are unrelated to this story (missing `ServiceConfiguration` and `App` exports targeted by Parity III Stories 4–6)

## Warp Conversation
https://app.warp.dev/conversation/045df0b1-0c0c-42fc-be5d-1acdd2fba698

Co-Authored-By: Oz <oz-agent@warp.dev>

### CODE REVIEW

File: tiferet/assets/error.py [hunk 1]
```diff
@@ -0,0 +1,685 @@
...
+    TOML_FILE_LOAD_ERROR_ID,
+    TOML_FILE_NOT_FOUND_ID,
+    UNSUPPORTED_CONFIG_FILE_TYPE_ID,
+    YAML_FILE_LOAD_ERROR_ID,
+    YAML_FILE_NOT_FOUND_ID,
+    YAML_FILE_SAVE_ERROR_ID,
+    create_default_error,
+)
+
+# *** constants (errors)
```
Comment (Ashatz) [inline tiferet/assets/error.py:89]: **Section header names deviate from TRD §4.6.**

TRD §4.6 calls for a single `# *** constants` section containing both the per-error definitions and the `DEFAULT_ERRORS` catalog dict (with `# ** constant: default_errors` as the dict's own label). The branch splits this into `# *** constants (errors)` here and `# *** constants (groups)` at line 608.

Fix: rename this header to `# *** constants` and drop the `# *** constants (groups)` header at line 608. The `# ** constant: default_errors` label on the dict is correct and stays.

File: tiferet/assets/error.py [hunk 2]
```diff
@@ -0,0 +1,685 @@
...
+    SQLITE_CONN_NOT_INITIALIZED_ID,
+    'SQLite Connection Not Initialized',
+    [(EN_US, 'SQLite connection not initialized. Must be used within a "with" block.')],
+)
+
+# ** constant: context_not_found
+CONTEXT_NOT_FOUND = create_default_error(
+    CONTEXT_NOT_FOUND_ID,
+    'Context Not Found',
+    [(EN_US, 'No context registered for domain type {domain_type}.')],
```
Comment (Ashatz) [inline tiferet/assets/error.py]: **`CONTEXT_NOT_FOUND` message text is missing a colon.**

`v2.x-proto` reads:
```
'No context registered for domain type: {domain_type}.'
```
The branch drops the colon after `type`. Since this is a new constant with no prior call-site on `main`, the proto wording is the authority. Add the colon: `'No context registered for domain type: {domain_type}.'`

**Review by Ashatz — COMMENTED** [review]
Review of #895 (Assets – Default Error Catalog Extraction and Constants Normalization) against the TRD and `v2.x-proto`.

**Overall:** The implementation is solid — correct scope, 73 identifiers (deduplication verified, 8 new IDs present), factory signature, catalog assembly, alphabetically-ordered imports in `error.py`, test coverage for all 4 required cases, and `__init__.py` re-export all match the TRD. Two minor deviations found.

**Branch-ahead notes (no action required):** `__init__.py` drops the proto's `__all__`, `RaiseError`, and sub-module re-exports (`core`, `app`, `feat`, `cli`, etc.) — these are Parity V (Stories 13/18) scope and are correctly excluded here. The flat `DEFAULT_ERRORS` dict vs. the proto's `CORE_DEFAULT_ERRORS` / `ADMIN_DEFAULT_ERRORS` / `SQLITE_DEFAULT_ERRORS` etc. sub-catalog split is also Parity V work; the TRD explicitly calls for a single flat dict. Proto entries for `APP_SESSION_NOT_FOUND_ID` and `INVALID_APP_SESSION_TYPE_ID` are absent — correctly excluded (Parity III Story 5a scope).

Reviewed by Oz — Co-Authored-By: Oz <oz-agent@warp.dev>
Conversation run ID: `019f9085-0732-7781-ab7a-a1c7bf232c46`

**Comment by Ashatz** [comment]
**Review follow-up — addressing inline comment #2**

Fixed: `CONTEXT_NOT_FOUND` message text updated to match `v2.x-proto` wording (colon added after "type").

Finding #1 (section headers `# *** constants (errors)` / `# *** constants (groups)` vs TRD's `# *** constants`) is accepted as-is per author review. Noted as a minor style deviation for the Collaboration Report.

Conversation link: https://app.warp.dev/conversation/019f9085-0732-7781-ab7a-a1c7bf232c46

Co-Authored-By: Oz <oz-agent@warp.dev>
